src/sentry/search/events/datasets/functions.py

```diff
@@ -0,0 +1,359 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from enum import Enum
+from typing import Callable, Mapping, Optional, Union
+
+from snuba_sdk import Column as SnColumn
+from snuba_sdk import Condition, Direction, Op, OrderBy
+
+from sentry.api.event_search import SearchFilter
+from sentry.search.events import builder
+from sentry.search.events.constants import EQUALITY_OPERATORS, PROJECT_ALIAS, PROJECT_NAME_ALIAS
+from sentry.search.events.datasets import field_aliases, filter_aliases
+from sentry.search.events.datasets.base import DatasetConfig
+from sentry.search.events.fields import (
+    ColumnArg,
+    Combinator,
+    Function,
+    InvalidFunctionArgument,
+    InvalidSearchQuery,
+    NumberRange,
+    NumericColumn,
+    SnQLFunction,
+    with_default,
+)
+from sentry.search.events.types import NormalizedArg, ParamsType, SelectType, WhereType
+
+
+class Kind(Enum):
+    DATE = "date"
+    DURATION = "duration"
+    INTEGER = "integer"
+    NUMBER = "number"
+    STRING = "string"
+
+
+class Duration(Enum):
+    NANOSECOND = "nanosecond"
+    MICROSECOND = "microsecond"
+    MILLISECOND = "millisecond"
+    SECOND = "second"
+    MINUTE = "minute"
+    HOUR = "hour"
+    DAY = "day"
+    WEEK = "week"
+
+
+# The only units available right now are duration based
+Unit = Duration
+
+
+@dataclass(frozen=True)
+class Column:
+
+    # the internal name in snuba
+    column: str
+    # type kind/type associated with this column
+    kind: Kind
+    # the external name to expose
+    alias: Optional[str] = None
+    # some kinds will have an unit associated with it
+    unit: Optional[Unit] = None
+
+
+COLUMNS = [
+    Column(alias="project.id", column="project_id", kind=Kind.INTEGER),
+    Column(alias="project_id", column="project_id", kind=Kind.INTEGER),
+    Column(alias="transaction", column="transaction_name", kind=Kind.STRING),
+    Column(alias="timestamp", column="timestamp", kind=Kind.DATE),
+    Column(alias="depth", column="depth", kind=Kind.INTEGER),
+    Column(alias="parent_fingerprint", column="parent_fingerprint", kind=Kind.INTEGER),
+    Column(alias="fingerprint", column="fingerprint", kind=Kind.INTEGER),
+    Column(alias="name", column="name", kind=Kind.STRING),
+    Column(alias="package", column="package", kind=Kind.STRING),
+    Column(alias="path", column="path", kind=Kind.STRING),
+    Column(alias="is_application", column="is_application", kind=Kind.INTEGER),
+    Column(alias="platform.name", column="platform", kind=Kind.STRING),
+    Column(alias="environment", column="environment", kind=Kind.STRING),
+    Column(alias="release", column="release", kind=Kind.STRING),
+    Column(alias="os.name", column="os_name", kind=Kind.STRING),
+    Column(alias="os.version", column="os_version", kind=Kind.STRING),
+    Column(alias="retention_days", column="retention_days", kind=Kind.INTEGER),
+    Column(alias="percentiles", column="percentiles", kind=Kind.DURATION),
```

Comment:
Does `percentiles` need to be column?

Reply:
yea, i was getting errors because we have this `NumericColumn` class that asserts against the column 

![image](https://user-images.githubusercontent.com/7349258/221603749-52b69951-6ffc-4b10-b6e1-ca54ef7e3a56.png)


if you ask me there's too many fancy things going on here, but my will to change it is quite low. i'm expecting to have to return once i start connecting the frontend. i can tackle other changes in a future PR